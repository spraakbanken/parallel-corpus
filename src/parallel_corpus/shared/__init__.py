"""Utilities."""

import re
from typing import TypeVar

from . import diffs

__all__ = ["diffs"]


ENDING_WHITESPACE = re.compile(r"\s$")


def end_with_space(s: str) -> str:
    if not s:
        return s
    return f"{s} " if (ENDING_WHITESPACE.fullmatch(s[-1]) is None) else s


def uniq(xs: list[str]) -> list[str]:
    used: set[str] = set()
    return [x for x in xs if x not in used and (used.add(x) or True)]  # type: ignore [func-returns-value]


A = TypeVar("A")
B = TypeVar("B")
