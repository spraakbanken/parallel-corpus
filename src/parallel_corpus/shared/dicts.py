"""dicts."""

from typing import TYPE_CHECKING, Callable, TypeVar

if TYPE_CHECKING:
    from _typeshed import SupportsRichComparison

    K = TypeVar("K", bound=SupportsRichComparison)  # type: ignore [syntax]
else:
    K = TypeVar("K")

A = TypeVar("A")
B = TypeVar("B")
V = TypeVar("V")


def modify(x: dict[K, V], k: K, default: V, f: Callable[[V], V]) -> V:  # noqa: D103
    x[k] = f(x.get(k) or default)
    return x[k]


def traverse(x: dict[K, A], k: Callable[[A, K], B], *, sort_keys: bool = False) -> list[B]:  # noqa: D103
    ks = sorted(x.keys()) if sort_keys else x.keys()
    return [k(x[i], i) for i in ks]


def filter_dict(x: dict[K, A], k: Callable[[A, K], bool]) -> dict[K, A]:  # noqa: D103
    return {id_: a for id_, a in x.items() if k(a, id_)}
