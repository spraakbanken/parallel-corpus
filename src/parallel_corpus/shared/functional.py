"""Functional utilities."""

from collections.abc import Sequence
from typing import Callable, TypeVar

A = TypeVar("A")


def take_last_while(predicate: Callable[[A], bool], xs: Sequence[A]) -> Sequence[A]:  # noqa: D103
    start = 0
    for e in reversed(xs):
        if not predicate(e):
            break
        start -= 1
    return xs[start:] if start < 0 else xs[:0]
