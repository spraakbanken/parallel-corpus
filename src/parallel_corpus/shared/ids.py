"""Ids."""

import re
from collections.abc import Iterable

DIGITS = re.compile(r"\d+")


def next_id(xs: Iterable[str]) -> int:
    """Calculate the next id to use from these identifiers.

    next_id([]) // => 0
    next_id(['t1', 't2', 't3']) // => 4
    next_id(['u2v5k1', 'b3', 'a0']) // => 6
    next_id(['77j66']) // => 78

    """
    curr_max = -1
    for x in xs:
        for digit in DIGITS.finditer(x):
            curr_max = max(curr_max, int(digit[0]))
    # xs.forEach(x => (x.match(/\d+/g) || []).forEach(i => (max = Math.max(max, parseInt(i)))))
    return curr_max + 1
