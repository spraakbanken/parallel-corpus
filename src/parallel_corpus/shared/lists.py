"""List."""

import copy
from typing import List, Tuple, TypeVar

A = TypeVar("A")


def rearrange(xs: List[A], begin: int, end: int, dest: int) -> List[A]:
    """Move a slice of the items and puts back them at some destination.

    rearrange([0, 1, 2, 3], 1, 2, 0) // => [1, 2, 0, 3]
    rearrange([0, 1, 2, 3], 1, 2, 3) // => [0, 3, 1, 2]

    rearrange([0, 1, 2, 3], 1, 2, 1) // => [0, 1, 2, 3]
    rearrange([0, 1, 2, 3], 1, 2, 2) // => [0, 1, 2, 3]
    """
    a, mid, z = split_at_3(xs, begin, end + 1)
    w = end - begin
    if dest > begin:
        dest -= w
    pre, post = split_at(a + z, dest)
    return pre + mid + post


def splice(xs: List[A], start: int, count: int, *insert) -> Tuple[List[A], List[A]]:  # noqa: ANN002, D103
    ys = copy.deepcopy(xs)
    zs = ys[start : (start + count)]
    ys[start : (start + count)] = insert
    return ys, zs


def split_at_3(xs: List[A], start: int, end: int) -> Tuple[List[A], List[A], List[A]]:
    """Split an array into three pieces.

    splitAt3('0123456'.split(''), 2, 4).map(xs => xs.join('')) // => ['01', '23', '456']
    splitAt3('0123456'.split(''), 2, 2).map(xs => xs.join('')) // => ['01', '', '23456']
    splitAt3('0123456'.split(''), 2, 9).map(xs => xs.join('')) // => ['01', '23456', '']
    splitAt3('0123456'.split(''), 0, 2).map(xs => xs.join('')) // => ['', '01', '23456']
    """
    ab, c = split_at(xs, end)
    a, b = split_at(ab, start)
    return a, b, c


def split_at(xs: List[A], index: int) -> Tuple[List[A], List[A]]:  # noqa: D103
    return xs[:index], xs[index:]
