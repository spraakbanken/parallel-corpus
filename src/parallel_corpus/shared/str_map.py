"""str_map."""

from typing import Callable, List, TypeVar

A = TypeVar("A")


def str_map(s: str, f: Callable[[str, int], A]) -> List[A]:  # noqa: D103
    return [f(s[i], i) for i in range(len(s))]
