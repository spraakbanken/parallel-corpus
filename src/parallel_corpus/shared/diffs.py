"""Diffs."""

import enum
from collections.abc import Generator
from itertools import starmap
from typing import Any, Callable, Generic, Optional, TypeVar, Union

import diff_match_patch as dmp_module
from typing_extensions import Self

from parallel_corpus.shared.str_map import str_map

dmp = dmp_module.diff_match_patch()

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")


class ChangeType(enum.IntEnum):  # noqa: D101
    DELETED = -1
    CONSTANT = 0
    INSERTED = 1


class Change(Generic[A, B]):  # noqa: D101
    def __init__(self, change: ChangeType, a: Optional[A] = None, b: Optional[B] = None) -> None:  # noqa: D107
        if change == ChangeType.DELETED and a is None:
            raise ValueError("`a` must be given for DELETED")
        if change == ChangeType.CONSTANT and (a is None or b is None):
            raise ValueError("both `a` and `b` must be given for CONSTANT")
        if change == ChangeType.INSERTED and b is None:
            raise ValueError("`b` must be given for INSERTED")
        self.change = change
        self.a = a
        self.b = b

    @classmethod
    def constant(cls, a: A, b: B) -> Self:  # noqa: D102
        return cls(ChangeType.CONSTANT, a=a, b=b)

    @classmethod
    def deleted(cls, a: A) -> Self:  # noqa: D102
        return cls(ChangeType.DELETED, a=a)

    @classmethod
    def inserted(cls, b: B) -> Self:  # noqa: D102
        return cls(ChangeType.INSERTED, b=b)

    def model_dump(self) -> dict[str, Union[int, A, B]]:  # noqa: D102
        out: dict[str, Union[int, A, B]] = {
            "change": int(self.change),
        }
        if self.a is not None:
            out["a"] = self.a
        if self.b is not None:
            out["b"] = self.b
        return out

    def __eq__(self, other: Any) -> bool:  # noqa: D105
        if not isinstance(other, Change):
            return NotImplemented
        return self.change == other.change and self.a == other.a and self.b == other.b

    def __hash__(self) -> int:  # noqa: D105
        return hash((self.change, self.a, self.b))

    def __repr__(self) -> str:  # noqa: D105
        return f"Change(change={self.change!r},a={self.a!r},b={self.b!r})"

    def __str__(self) -> str:  # noqa: D105
        return f"Change(change={self.change},a={self.a},b={self.b})"


def char_stream() -> Generator[str, None, None]:
    """Make a stream of all unicode characters.

    We need this because the diff-match-patch library is hard-coded to work on characters.

    To make a polymorphic diff each unique element is assigned a unique character.
    We translate them back to the opaque type after diffing via the characters.
    This is used in `hdiff`.

    >>> chars = char_stream()
    >>> assert ord(next(chars)) == 0
    >>> assert ord(next(chars)) == 1
    >>> assert ord(next(chars)) == 2
    >>> assert ord(next(chars)) == 3

    """
    i = 0
    while True:
        yield chr(int(str(i), base=16))
        i += 1


def hdiff(  # noqa: D103
    xs: list[A],
    ys: list[B],
    a_cmp: Callable[[A], str] = str,
    b_cmp: Callable[[B], str] = str,
) -> list[Change[A, B]]:
    to: dict[str, str] = {}
    a_from: dict[str, list[A]] = {}
    b_from: dict[str, list[B]] = {}
    chars = char_stream()

    def assign(c: C, c_cmp: Callable[[C], str], c_from: dict[str, list[C]]) -> str:
        s = c_cmp(c)
        u = to.get(s)
        if u is None:
            u = next(chars)
            to[s] = u
        arr = c_from.get(u)
        if not arr:
            arr = []
            c_from[u] = arr
        arr.append(c)
        return u

    s1 = "".join(assign(a, a_cmp, a_from) for a in xs)
    s2 = "".join(assign(b, b_cmp, b_from) for b in ys)
    d = dmp.diff_main(s1, s2)

    def str_map_change(change: int) -> Callable[[str, int], Change]:
        def inner(c: str, _: int) -> Change:
            if change == 0:
                a = a_from.get(c, []).pop(0)
                b = b_from.get(c, []).pop(0)
                return Change.constant(a, b)
            if change == -1:
                a = a_from.get(c, []).pop(0)
                return Change.deleted(a)
            if change == 1:
                b = b_from.get(c, []).pop(0)
                return Change.inserted(b)
            raise RuntimeError("diff-match-patch change not in range [-1,1]")

        return inner

    def map_change(change: int, cs):  # noqa: ANN001, ANN202
        return str_map(cs, str_map_change(change))

    out = []
    for changes in starmap(map_change, d):
        # print(f"{changes=}")
        out.extend(changes)
    return out


def token_diff(s1: str, s2: str) -> list[tuple[int, str]]:  # noqa: D103
    d = dmp.diff_main(s1, s2)
    dmp.diff_cleanupSemantic(d)
    return d
