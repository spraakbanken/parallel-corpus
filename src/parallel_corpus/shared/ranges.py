"""Ranges."""

import itertools
from typing import TypedDict

from parallel_corpus.shared.diffs import token_diff
from parallel_corpus.shared.functional import take_last_while

EditRange = TypedDict("EditRange", {"from": int, "to": int, "insert": str})


def edit_range(s0: str, s: str) -> EditRange:
    """Create an EditRange.

    >>> edit_range('0123456789', '0189')
    {'from': 2, 'to': 8, 'insert': ''}

    >>> edit_range('0123456789', '01')
    {'from': 2, 'to': 10, 'insert': ''}

    >>> edit_range('0123456789', '89')
    {'from': 0, 'to': 8, 'insert': ''}

    >>> edit_range('0123456789', '')
    {'from': 0, 'to': 10, 'insert': ''}

    >>> edit_range('0123456789', '01xyz89')
    {'from': 2, 'to': 8, 'insert': 'xyz'}

    >>> edit_range('0123456789', '01xyz')
    {'from': 2, 'to': 10, 'insert': 'xyz'}

    >>> edit_range('0123456789', 'xyz89')
    {'from': 0, 'to': 8, 'insert': 'xyz'}

    >>> edit_range('0123456789', 'xyz')
    {'from': 0, 'to': 10, 'insert': 'xyz'}

    >>> edit_range('', '01')
    {'from': 0, 'to': 0, 'insert': '01'}
    """
    patches = token_diff(s0, s)
    pre = list(itertools.takewhile(lambda i: i[0] == 0, patches))
    post = take_last_while(lambda i: i[0] == 0, patches)
    from_ = len("".join(i[1] for i in pre))
    postlen = len("".join(i[1] for i in post))
    to = len(s0) - postlen
    insert = s[from_ : (len(s) - (len(s0) - to))]
    return {"from": from_, "to": to, "insert": insert}
