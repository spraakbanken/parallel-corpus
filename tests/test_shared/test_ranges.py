import string

import pytest

from parallel_corpus.shared.ranges import edit_range


@pytest.mark.parametrize(
    ("s0", "s"),
    [
        (string.digits, "0189"),
        (string.digits, "01"),
        (string.digits, "89"),
        (string.digits, ""),
        (string.digits, "01xyz89"),
        (string.digits, "01xyz"),
        (string.digits, "xyz89"),
        (string.digits, "xyz"),
        ("", "01"),
    ],
)
def test_edit_range(s0: str, s: str, snapshot) -> None:
    assert edit_range(s0, s) == snapshot
