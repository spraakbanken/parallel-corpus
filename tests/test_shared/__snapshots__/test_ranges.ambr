# serializer version: 1
# name: test_edit_range[-01]
  dict({
    'from': 0,
    'insert': '01',
    'to': 0,
  })
# ---
# name: test_edit_range[0123456789-0189]
  dict({
    'from': 2,
    'insert': '',
    'to': 8,
  })
# ---
# name: test_edit_range[0123456789-01]
  dict({
    'from': 2,
    'insert': '',
    'to': 10,
  })
# ---
# name: test_edit_range[0123456789-01xyz89]
  dict({
    'from': 2,
    'insert': 'xyz',
    'to': 8,
  })
# ---
# name: test_edit_range[0123456789-01xyz]
  dict({
    'from': 2,
    'insert': 'xyz',
    'to': 10,
  })
# ---
# name: test_edit_range[0123456789-89]
  dict({
    'from': 0,
    'insert': '',
    'to': 8,
  })
# ---
# name: test_edit_range[0123456789-]
  dict({
    'from': 0,
    'insert': '',
    'to': 10,
  })
# ---
# name: test_edit_range[0123456789-xyz89]
  dict({
    'from': 0,
    'insert': 'xyz',
    'to': 8,
  })
# ---
# name: test_edit_range[0123456789-xyz]
  dict({
    'from': 0,
    'insert': 'xyz',
    'to': 10,
  })
# ---
